cbmc_viewer/byteopt.py

```diff
@@ -0,0 +1,140 @@
+
+class ByteOpSummary:
+
+    def __init__(self, json_file, srcdir):
+
+        self.summary = do_make_byteop(json_file, srcdir)
+        self.validate()
+
+    def __str__(self):
+        """Render the byte op summary as html."""
+        return templates.render_byteop_summary(self.summary)
+
+    def validate(self):
+        """Validate members of a summary object."""
+
+        return voluptuous.humanize.validate_with_humanized_errors(
+            self.__dict__, VALID_BYTE_OP_SUMMARY
+        )
+
+    def dump(self, filename=None, outdir='.'):
+        """Write the byte op summary to a file rendered as html."""
+
+        util.dump(self, filename or "byteop.html", outdir)
+
+################################################################
+# Json key tags to access elements in cbmc json output
+
+JSON_EXTRACT_KEY = 'byteExtractStats'
+JSON_EXTRACT_LIST_KEY = 'byteExtractList'
+JSON_EXTRACT_COUNT_KEY = 'numOfExtracts'
+JSON_UPDATE_KEY = 'byteUpdateStats'
+JSON_UPDATE_LIST_KEY = 'byteUpdateList'
+JSON_UPDATE_COUNT_KEY = 'numOfUpdates'
+JSON_BYTE_OP_KEY = 'byteOpsStats'
+JSON_SOURCE_LOC_KEY = 'sourceLocation'
+
+################################################################
+# Utility functions to generate byteop summary
+
+def remove_ssa_expr(json_data):
+    JSON_REMOVE_KEY = 'ssaExpr'
+
+    json_extract_list = json_data.get(JSON_EXTRACT_KEY).get(JSON_EXTRACT_LIST_KEY)
+    json_update_list = json_data.get(JSON_UPDATE_KEY).get(JSON_UPDATE_LIST_KEY)
+
+    for item in json_extract_list:
+        item.pop(JSON_REMOVE_KEY)
+    for item in json_update_list:
+        item.pop(JSON_REMOVE_KEY)
+
+def get_src_location(json_data, root):
+    json_extract_list = json_data.get(JSON_EXTRACT_KEY).get(JSON_EXTRACT_LIST_KEY)
+    json_update_list = json_data.get(JSON_UPDATE_KEY).get(JSON_UPDATE_LIST_KEY)
+
+    root = srcloct.abspath(root)
+    for item in json_extract_list:
+        item[JSON_SOURCE_LOC_KEY] = srcloct.json_srcloc(item[JSON_SOURCE_LOC_KEY], root)
+    for item in json_update_list:
+        item[JSON_SOURCE_LOC_KEY] = srcloct.json_srcloc(item[JSON_SOURCE_LOC_KEY], root)
+
+def get_byte_op_metrics(json_file, root):
+    summary = {}
+
+    json_data = parse.parse_json_file(json_file)
+    for item in json_data:
+        if JSON_BYTE_OP_KEY in item.keys():
+            byteop = item[JSON_BYTE_OP_KEY]
+            remove_ssa_expr(byteop)
+            get_src_location(byteop, root)
+
+            summary.setdefault('byteExtractList',[]).extend(
+                byteop[JSON_EXTRACT_KEY][JSON_EXTRACT_LIST_KEY])
+            summary['numOfExtracts'] = summary.setdefault('numOfExtracts',0) + \
+                byteop[JSON_EXTRACT_KEY][JSON_EXTRACT_COUNT_KEY]
+            summary.setdefault('byteUpdateList',[]).extend(
+                byteop[JSON_UPDATE_KEY][JSON_UPDATE_LIST_KEY])
+            summary['numOfUpdates'] = summary.setdefault('numOfUpdates',0) + \
+                byteop[JSON_UPDATE_KEY][JSON_UPDATE_COUNT_KEY]
+
+    if not summary:
```

Comment:
I would set summary to a default value before the for loop, and update it in the for loop if the appropriate key is found.